Adds Reporter for instrumentation to use after they record a span

adriancole opened the pull request:
## Reporter
After recording an operation into a span, it needs to be reported out of process. There are two
builtin reporter implementations in this library, although you are free to create your own.

The simplest mechanism is printing out spans as they are reported.

```java
Reporter.CONSOLE.report(span);
```

Reporter is very similar to the following:
* Finagle's DeadlineSpanMap
* zipkin-finagle's SpanRecorder
* Brave's SpanCollector
* HTrace's ZipkinSpanReceiver
* Wingtips' ZipkinSpanSender
* Sleuth's ZipkinSpanReporter
* Jaeger's Reporter

and.. probably many others.

## AsyncReporter
AsyncReporter is how you actually get spans to zipkin. By default, it waits up to a second
before flushes any pending spans out of process via a Sender.

```java
reporter = AsyncReporter.builder(URLConnectionSender.create("http://localhost:9411/api/v1/spans"))
                        .build(Encoder.THRIFT_BYTES);

// Schedules the span to be sent, and won't block the calling thread on I/O
reporter.report(span);
```

adriancole commented:
need to make tests, etc (mostly port them from other places).

@nicmunroe lemme know what you think

adriancole commented:
@schlosna ik that you're on holiday, but probably you'll be interested in this. it is pre-1.0 so we can adjust as we learn

adriancole commented:
also cc @felixbarny

adriancole commented:
going to rejig batchreporter a bit so that pooled buffers (like netty or okio) are possible.

nicmunroe commented on core/src/main/java/zipkin/reporter/BatchReporter.java:
```diff
@@ -0,0 +1,209 @@
+  @Override
+  public void report(Span span) {
+    metrics.incrementSpans(1);
+    if (!pending.offer(spanEncoder.encode(span))) {
```
Any thought to alternate behavior besides simply dropping the span? You could have it immediately flush the sender with whatever you have even if it's not the regularly scheduled sending time. Maybe the ability to have it auto-flush like that when it's 80% full or something?

These alternate behaviors could add some non-trivial complexity, so maybe it's not worth it at this point (or ever depending on if people would actually use it). Something to think about though.

adriancole replied:
Good thoughts. So I'd expect the queue (or buffer) to always be bounded.
What do do when it is full (besides dropping) is a good question. There's
another matter which is whether to drop a span too large to send. For
example kafka has a max message size which people have (usually
erroneously) exceeded.

You are right that we can rejig things to allow a flush ahead of schedule
(based on size) though that would more of a precaution than a cure of a
full queue.

Lets revisit this again on next round.. thanks for the notes!

On 22 Aug 2016 10:17 am, "Nic Munroe" <notifications@github.com> wrote:

> In core/src/main/java/zipkin/reporter/BatchReporter.java
> <https://github.com/openzipkin/zipkin-reporter-java/pull/7#discussion_r75610994>
> :
>
> > +  }
> > +
> > +  BatchReporter(Builder builder, Logger logger) { // Visible for testing
> > +    this.sender = checkNotNull(builder.sender, "sender");
> > +    this.logger = logger;
> > +    this.spanEncoder = checkNotNull(builder.spanEncoder, "spanEncoder");
> > +    this.metrics = builder.metrics == null ? ReporterMetrics.NOOP_METRICS : builder.metrics;
> > +    this.pending = new LinkedBlockingDeque<>(builder.maxQueueDepth);
> > +    this.flusher = builder.flushInterval > 0 ? new Flusher(this, builder.flushInterval,
> > +        "Flusher(" + sender + ")") : null;
> > +  }
> > +
> > +  @Override
> > +  public void report(Span span) {
> > +    metrics.incrementSpans(1);
> > +    if (!pending.offer(spanEncoder.encode(span))) {
>
> Any thought to alternate behavior besides simply dropping the span? You
> could have it immediately flush the sender with whatever you have even if
> it's not the regularly scheduled sending time. Maybe the ability to have it
> auto-flush like that when it's 80% full or something?
>
> These alternate behaviors could add some non-trivial complexity, so maybe
> it's not worth it at this point (or ever depending on if people would
> actually use it). Something to think about though.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin-reporter-java/pull/7/files/cdafba588b33e42a6a8a148bb733dd0b9bae5806#r75610994>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD613UKlayASO7K7RnKI9yeJHVddg2Dks5qiQbFgaJpZM4JpQw9>
> .
>

nicmunroe commented:
@adriancole Looks good to me. Separating out the reporter vs. the sender makes a lot of sense and is the direction I was accidentally heading.

adriancole commented:
Added rough-cut of byte tracking and the ability to support non-bytearray buffers

adriancole commented:
before merging this, I'm going to practice with it by adding tests that bind it to at least htrace and brave. That will show how it works in real life.

adriancole commented:
just pushed a boatload of integration examples. I'm not currently happy with the api, just using the examples to help me drive down complexity (especially around building things which may or may not need to be transformed)

adriancole commented:
alright, so not happy with the api, yet, but getting there. obv need to backfill tests. Here are some notes:

* added BoundedQueue, which allows you to drain up to a certain size in bytes. It guards entry to ensure you can't put something larger than that on. This can be used to ask for up to N bytes to go to the sender at a time (as opposed to N spans). This only works when you can size the object (fine in json or thrift encoding).

* changed to htrace's draining approach. Instead of using a scheduler, this polls until there are spans available (up to 1 second as opposed to before which was minimum 1 second). This also controls the shutdown frequency. The difference now is that we report as soon as spans arrive (or as long as there's a backlog). we could pretty easily put a throttle on it later, if need be.

cc @nicmunroe don't pay too much attention to the details of the code, as I've another day of work cleaning it up. That said, lemme know if the approach and the api integration looks alright.

adriancole commented on core/src/main/java/zipkin/reporter/AsyncReporter.java:
```diff
@@ -0,0 +1,253 @@
+        } finally {
+          shutdown.countDown();
+        }
+      }, "Flusher(" + builder.sender + ")").start();
```
I know daemon threads are a polarizing topic.. if we make this daemon, then people could forget to call close and process shutdown will silently toss things.. If we don't make this daemon, someone who forgets to call close will have their process hung and even looking at this thread name.. maybe they will still complain.

Any thoughts?

adriancole commented on core/src/main/java/zipkin/reporter/internal/BoundedQueue.java:
```diff
@@ -0,0 +1,150 @@
+import zipkin.internal.Nullable;
+
+/** Special queue that can measure the size of entries as they are added. */
+public final class BoundedQueue<E> {
```
loosely based on ArrayBlockingQueue

adriancole commented on core/src/main/java/zipkin/reporter/internal/BoundedQueue.java:
```diff
@@ -0,0 +1,150 @@
+    Object toAdd = element;
+    if (byteCounter != null) {
+      int byteCount = byteCounter.byteCount(element);
+      if (byteCount > maxByteCount) return false;
```
returning false, here, but I could make this a numeric return like caffeine has. we could define one for full and another for over limit. throwing exception is possible, too, though probably not as cheap as a number.

Ex. here's caffeine's codes..
```java
interface Buffer<E> {
  int FULL = 1;
  int SUCCESS = 0;
  int FAILED = -1;
```

https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/Buffer.java#L35

felixbarny replied:
I'd prefer enums over int constants any time.

adriancole replied:
sgtm

adriancole commented on reporter/wingtips/src/main/java/zipkin/reporter/wingtips/WingtipsSpanReporter.java:
```diff
@@ -0,0 +1,51 @@
+  final AsyncReporter<Span> delegate;
+
+  public WingtipsSpanReporter(Sender<byte[]> sender) {
+    this.delegate = AsyncReporter.builder(sender).build(Encoder.THRIFT_BYTES);
```
doing sender as the first arg and encoder as a terminal arg helps a bit as it keeps the builder less complex. Not sure if this pattern is objectional or not..

adriancole replied:
one limitation is it prevents the "toBuilder()" pattern from working when we do it like this... I'm not terribly concerned about this as I don't think people will be boilerplating async reporters.

adriancole commented:
ps not obvious yet, but the components in this PR (particularly bounded queue) support a policy something like this, which can be enforced very early (ex as soon as the span is reported), vs now (very late.. like all the way at kafka!)

maxBufferedSpanBytes = 32MiB // currently unenforcable without this code
maxSpansBytesPerMessage = 1MiB // currently transport-specific

felixbarny commented on core/src/main/java/zipkin/reporter/internal/BoundedQueue.java:
```diff
@@ -0,0 +1,150 @@
+  public boolean offer(E element) {
+    Object toAdd = element;
+    if (byteCounter != null) {
+      int byteCount = byteCounter.byteCount(element);
```
Does that mean spans have to be serialized twice? Once to get the size and once more to actually send them over the wire?

adriancole replied:
nope. this just reports the size of the encoded buffer. ex in the case of a
byte array bytes.length

On Tue, Aug 23, 2016 at 11:14 PM, Felix <notifications@github.com> wrote:

> In core/src/main/java/zipkin/reporter/internal/BoundedQueue.java
> <https://github.com/openzipkin/zipkin-reporter-java/pull/7#discussion_r75887455>
> :
>
> > +
> > +    abstract int byteCount();
> > +  }
> > +
> > +  BoundedQueue(int size, int maxBatchSize, ByteCounter<E> byteCounter, int maxByteCount) {
> > +    this.items = new Object[size]; // cannot instantiate a generic type
> > +    this.maxBatchSize = maxBatchSize;
> > +    this.byteCounter = byteCounter;
> > +    this.maxByteCount = maxByteCount;
> > +  }
> > +
> > +  /** Returns true if the element could be added or false if the queue is full */
> > +  public boolean offer(E element) {
> > +    Object toAdd = element;
> > +    if (byteCounter != null) {
> > +      int byteCount = byteCounter.byteCount(element);
>
> Does that mean spans have to be serialized twice? Once to get the size and
> once more to actually send them over the wire?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin-reporter-java/pull/7/files/a4fd2cf9406f13fe83d31a3fce7603fff0e39fb6#r75887455>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD619jt1hGsGJzLyEmav31qI5CGXO8xks5qiw5EgaJpZM4JpQw9>
> .
>

felixbarny replied:
But as far as I understand, `E element` is an instance of `Span` and not a `byte[]`.

felixbarny replied:
Ok, I see `E` is a `byte[]`: `URLConnectionSender<byte[]> create(String endpoint)`

adriancole replied:
So this type requires a lambda `ByteCounter` to be passed, to count the size of `E` (there's no notion of span in this type). We bind that to `Encoder::byteCount`, which usually binds `E` to `byte[]`. (ex as `Encoder.THRIFT_BYTES` does).

However, some will not like byte arrays thrashing, so they might bind `E` to an `okio.Buffer`, a `ByteCounter` for that would return its available bytes.

Thanks for asking the question.. it will help me write javadocs.

adriancole commented on core/src/main/java/zipkin/reporter/AsyncReporter.java:
```diff
@@ -0,0 +1,253 @@
+  @AutoValue
+  static abstract class EncodingAsyncReporter<S, B> extends AsyncReporter<S> {
+
+    abstract Encoder<S, B> encoder();
```
@felixbarny here's where I'd add javadoc explaining that this thing ends up controlling the element type of the bounded queue, using examples of `S` being `zipkin.Span` and `B` being `byte[]`.

adriancole commented on core/src/main/java/zipkin/reporter/Sender.java:
```diff
@@ -15,41 +15,45 @@
  */
-// @FunctionalInterface
-public interface Sender extends Component {
+public interface Sender<S> extends Component {
```
self-note: change this to B!

nicmunroe commented:
@adriancole LGTM so far. The simplicity of the example `WingtipsSpanReporter` you created tells me it's good. Proper separation of concerns with interfaces defining the contracts leads to easy integrations. :+1: 

> cc @nicmunroe don't pay too much attention to the details of the code, as I've another day of work cleaning it up. That said, lemme know if the approach and the api integration looks alright.

adriancole commented:
going to do a big refactor again. by removing the ability to create a non-encoding reporter, the whole codebase is simpler (ex makes things like what's a buffer or not extremely obvious).

after that, I'll focus on ensuring messages are possible to send. For example, we often have issues due to kafka `message.max.bytes` where producers inadvertently send more than possible to consume. This happens with scribe, too, just less frequently.

adriancole commented:
ok code is now simplified a lot. There's only one type of constraint for sending, which is the total bytes per message. This is much more appropriate for reporters as they don't care what is inside most of the time.

adriancole commented:
OK reasonably happy with the api at this point.

* sender decides how much it can accept (messageMaxBytes)
* reporter honors that by not dequeuing more than this at a time

Now's details time.. mostly backfilling tests

adriancole commented:
Added first tests (also added the ability to control worker count as that exists in htrace)

adriancole commented:
pulled sender-specific changes into a separate PR as they are relatively stable:  https://github.com/openzipkin/zipkin-reporter-java/pull/8

adriancole commented:
After benchmarking, I'm confident in the async approach. With default serialization overhead, `AsyncReporter.report(Span)` takes about 2microseconds on my laptop. I've tested that even with high contention (8 threads reporting), no spans are dropped and the reporting throughput is linear. 

I'm going to backfill more tests and plan to merge in the next day or two. fyi @nicmunroe @felixbarny

```
Benchmark                                                        Mode  Cnt      Score     Error   Units
AsyncReporterBenchmarks.high_contention                         thrpt   15   1550.877 ± 128.986  ops/ms
AsyncReporterBenchmarks.high_contention:high_contention_flush   thrpt   15    158.727 ±  13.241  ops/ms
AsyncReporterBenchmarks.high_contention:high_contention_report  thrpt   15   1392.150 ± 115.747  ops/ms
AsyncReporterBenchmarks.high_contention:messagesDropped         thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.high_contention:messagesSent            thrpt   15    158.745 ±  13.234  ops/ms
AsyncReporterBenchmarks.high_contention:spansDropped            thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.high_contention:spansReported           thrpt   15  11139.567 ± 924.662  ops/ms
AsyncReporterBenchmarks.mild_contention                         thrpt   15   1143.694 ±  21.488  ops/ms
AsyncReporterBenchmarks.mild_contention:messagesDropped         thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.mild_contention:messagesSent            thrpt   15    172.181 ±   4.248  ops/ms
AsyncReporterBenchmarks.mild_contention:mild_contention_flush   thrpt   15    172.179 ±   4.248  ops/ms
AsyncReporterBenchmarks.mild_contention:mild_contention_report  thrpt   15    971.515 ±  17.951  ops/ms
AsyncReporterBenchmarks.mild_contention:spansDropped            thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.mild_contention:spansReported           thrpt   15   1943.060 ±  35.909  ops/ms
AsyncReporterBenchmarks.no_contention                           thrpt   15    734.925 ±  25.125  ops/ms
AsyncReporterBenchmarks.no_contention:messagesDropped           thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.no_contention:messagesSent              thrpt   15    183.372 ±  15.215  ops/ms
AsyncReporterBenchmarks.no_contention:no_contention_flush       thrpt   15    183.372 ±  15.215  ops/ms
AsyncReporterBenchmarks.no_contention:no_contention_report      thrpt   15    551.553 ±  17.670  ops/ms
AsyncReporterBenchmarks.no_contention:spansDropped              thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.no_contention:spansReported             thrpt   15    551.557 ±  17.669  ops/ms
```

adriancole commented:
Measured that there's no notable backlog even at very high reporting speeds (10spans/us which no zipkin system can consume). This means we won't ever need more than one thread to invoke senders. However, we do need to be careful about senders who aren't asynchronous as they can block the sender loop. I made a note of this in javadoc and this is something we can followup with on a per-sender basis.

```
Benchmark                                                 Mode  Cnt      Score     Error   Units
AsyncReporterBenchmarks.high_contention                  thrpt   15   1252.066 ±  89.663  ops/ms
AsyncReporterBenchmarks.high_contention:messages         thrpt   15      2.795 ±   0.200  ops/ms
AsyncReporterBenchmarks.high_contention:messagesDropped  thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.high_contention:spanBacklog      thrpt   15      0.002 ±   0.009  ops/ms
AsyncReporterBenchmarks.high_contention:spans            thrpt   15  10017.368 ± 717.519  ops/ms
AsyncReporterBenchmarks.high_contention:spansDropped     thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.mild_contention                  thrpt   15    933.658 ±  62.117  ops/ms
AsyncReporterBenchmarks.mild_contention:messages         thrpt   15      0.521 ±   0.035  ops/ms
AsyncReporterBenchmarks.mild_contention:messagesDropped  thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.mild_contention:spanBacklog      thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.mild_contention:spans            thrpt   15   1867.322 ± 124.236  ops/ms
AsyncReporterBenchmarks.mild_contention:spansDropped     thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.no_contention                    thrpt   15    551.589 ±  41.728  ops/ms
AsyncReporterBenchmarks.no_contention:messages           thrpt   15      0.154 ±   0.012  ops/ms
AsyncReporterBenchmarks.no_contention:messagesDropped    thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.no_contention:spanBacklog        thrpt   15        ≈ 0            ops/ms
AsyncReporterBenchmarks.no_contention:spans              thrpt   15    551.589 ±  41.728  ops/ms
AsyncReporterBenchmarks.no_contention:spansDropped       thrpt   15        ≈ 0            ops/ms
```
